### PR TITLE
Added extension with IConfiguration functionality to IReloadingManage…

### PR DESCRIPTION
**Pull request by starkmsu**
…r for logging

### CODE REVIEW

File: src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                if (wasReloaded)
+                    foreach (var callbackPair in _callbacks)
+                    {
+                        try
+                        {
+                            callbackPair.Item1(callbackPair.Item2);
+                        }
+                        catch (Exception ex)
+                        {
+                            Console.WriteLine(ex);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs]: May be will be better to use ILog?

File: src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        /// </summary>
+        /// <param name="callback">The System.Action`1 to invoke.</param>
+        /// <param name="state">State to be passed into the callback.</param>
+        /// <returns>An System.IDisposable that is used to unregister the callback.</returns>
+        public IDisposable RegisterChangeCallback(Action<object> callback, object state)
+        {
+            var callBackPair = (callback, state);
+            _callbacks.Add(callBackPair);
+
+            return new CallbackUnregisterHandler(() => _callbacks.Remove(callBackPair));
```
Comment (KonstantinRyazantsev) [inline src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs]: Is it one time callback? If so, will be nice to note this in summary and/or method name

File: src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsConfigurationSection.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+using Microsoft.Extensions.Primitives;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Lykke.SettingsReader.ReloadingManager.Configuration
+{
+    /// <summary>
+    /// Microsoft.Extensions.Configuration.IConfigurationSection implementation for logging configuration section
+    /// </summary>
+    public sealed class SettingsConfigurationSection<T> : IConfigurationSection
```
Comment (KonstantinRyazantsev) [inline src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsConfigurationSection.cs]: Could it be internal?

File: src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Primitives;
+
+namespace Lykke.SettingsReader.ReloadingManager.Configuration
+{
+    /// <summary>
+    /// Provides Microsoft.Extensions.Primitives.IChangeToken functionality for LoggingConfiguration
+    /// </summary>
+    public class SettingsChangeToken<T> : IChangeToken
```
Comment (KonstantinRyazantsev) [inline src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs]: Could it be internal?

File: src/Lykke.SettingsReader/ReloadingManager/LocalSettingsReloadingManager.cs [hunk 5]
```diff
@@ -1,10 +1,11 @@
 ﻿using System;
 using System.IO;
 using System.Threading.Tasks;
+using Lykke.SettingsReader.ReloadingManager.Configuration;
 
 namespace Lykke.SettingsReader
 {
-    public class LocalSettingsReloadingManager<TSettings> : ReloadingManagerBase<TSettings>
+    public class LocalSettingsReloadingManager<TSettings> : ReloadingManagerWithConfigurationBase<TSettings>
```
Comment (KonstantinRyazantsev) [inline src/Lykke.SettingsReader/ReloadingManager/LocalSettingsReloadingManager.cs]: Will be nice to mark this class as [PublicAPI], as well as anothe public classes/interfaces

File: src/Lykke.SettingsReader/ReloadingManager/Configuration/IReloadingManagerWithConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    /// <summary>
+    /// Interface for IReloadingManager that provides Microsoft.Extensions.Configuration.IConfiguration functionality
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public interface IReloadingManagerWithConfiguration<T> : IReloadingManager<T>
+    {
+        /// <summary>
+        /// Property that contains Microsoft.Extensions.Configuration.IConfiguration implementation for IReloadingManager
+        /// </summary>
+        IConfiguration SettingsConfiguration { get; }
```
Comment (KonstantinRyazantsev) [inline src/Lykke.SettingsReader/ReloadingManager/Configuration/IReloadingManagerWithConfiguration.cs]: Will be nice to mark all public properties, method return values and method parameters, which types are reference-types, with [NotNull] or [CanBeNull]

Comment (starkmsu) [reply src/Lykke.SettingsReader/ReloadingManager/Configuration/IReloadingManagerWithConfiguration.cs]: I decided not to add new dependency to JetBrains nuget for this

File: src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs (hunk 2 above)
Comment (starkmsu) [reply src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsChangeToken.cs]: The sole purpose of CallbackUnregisterHandler is to call unregister callback via Dispose call. So I'm not sure what should be added in summary. Method's signature can't be change cause it's a part of IChangeToken interface.

File: src/Lykke.SettingsReader/ReloadingManager/LocalSettingsReloadingManager.cs (hunk 5 above)
Comment (starkmsu) [reply src/Lykke.SettingsReader/ReloadingManager/LocalSettingsReloadingManager.cs]: Same as for NulNull - decided not to add a new reference to JetBrains nuget

File: src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsConfigurationSection.cs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            IReloadingManager<T> manager,
+            string key,
+            string path,
+            JToken token)
+        {
+            _manager = manager;
+            Key = key;
+            Path = path;
+            _token = token;
+            if (token == null)
```
Comment (MikhailGerasimov) [inline src/Lykke.SettingsReader/ReloadingManager/Configuration/SettingsConfigurationSection.cs]: Pattern matching would be clear here

**Pull request merged**